src/python/lib/ensembl/compara/db/dbconnection.py

```diff
@@ -0,0 +1,147 @@
+URL = TypeVar('URL', str, sqlalchemy.engine.url.URL)
+
+
+class DBConnection:
+    """Database connection handler, providing also the database's schema and properties.
+
+    Args:
+        url: URL to the database, e.g. ``mysql://user:passwd@host:port/my_db``.
+
+    """
+    def __init__(self, url: URL) -> None:
+        self._engine = create_engine(url)
+        self._load_metadata()
+
+    def _load_metadata(self) -> None:
+        """Loads the metadata information of the database."""
+        # Just reflect() is not enough as it would not delete tables that no longer exist
+        self._metadata = sqlalchemy.MetaData(bind=self._engine)
+        self._metadata.reflect()
+
+    @property
+    def url(self) -> str:
+        """Returns the database URL."""
+        return str(self._engine.url)
+
+    @property
+    def db_name(self) -> str:
+        """Returns the database name."""
+        return self._engine.url.database
+
+    @property
+    def host(self) -> str:
+        """Returns the database host."""
+        return self._engine.url.host
+
+    @property
+    def tables(self) -> Dict[str, sqlalchemy.schema.Table]:
+        """Returns a dictionary of :class:`~sqlalchemy.schema.Table` objects keyed to their name."""
+        return self._metadata.tables
+
+    def get_primary_key_columns(self, table: str) -> List[str]:
+        """Returns the list of primary key column names for the given table.
+
+        Args:
+            table: Table name.
+
+        """
+        return [col.name for col in self.tables[table].primary_key.columns.values()]
+
+    def get_columns(self, table: str) -> List[str]:
+        """Returns the list of column names for the given table.
+
+        Args:
+            table: Table name.
+
+        """
+        return [col.name for col in self.tables[table].columns]
+
+    @property
+    def schema_type(self) -> str:
+        """Returns the schema type of the database, located in the ``meta`` table."""
+        result = self.execute(
+            select([self.tables['meta'].columns.meta_value]).where(text('meta_key = "schema_type"'))
+        )
+        return result.first()[0]
+
+    @property
+    def schema_version(self) -> int:
+        """Returns the schema version of the database, located in the ``meta`` table."""
+        result = self.execute(
+            select([self.tables['meta'].columns.meta_value]).where(text('meta_key = "schema_version"'))
+        )
+        return int(result.first()[0])
+
+    def connect(self) -> sqlalchemy.engine.Connection:
+        """Returns a new :class:`~sqlalchemy.engine.Connection` object."""
+        return self._engine.connect()
+
+    @contextlib.contextmanager
+    def begin(self, *args) -> sqlalchemy.engine.Connection:
+        """Returns a context manager delivering a :class:`~sqlalchemy.engine.Connection` with a
+        :class:`~sqlalchemy.engine.Transaction` established.
+        """
+        with self._engine.begin(*args) as conn:
+            yield conn
```

Comment:
How does this differ from `session_scope` ?

Reply:
This returns a `connection` object without any type of commit/rollback logic. The other one returns a `session` object, and has the commit/rollback logic. In `sqlalchemy` you can see them used for different purposes. As I understand it, in general terms, you use `connection` if you don't have a map of the tables, use `session` if you do.

Reply:
So that is why I couldn't find the equivalent of the session code for connections (I searched for it). Cheers, knowing that is quite useful :+1: 
Yes, that is my understanding after all this time reading and looking for SQLAlchemy ways of doing things. The community seems to be more into the ORM side as they claim is more pythonic :man_shrugging:
I wasn't sure so that is why I did it that way. I will test your proposal and if it works, I will commit the change :+1: